### PR TITLE
[MetriCam2] Image cache

### PR DESCRIPTION
**Pull request by jangernert**

### CODE REVIEW

**Comment by jangernert** [comment]
@f00f 
Just something I noticed with the MvBlueSirius Implementation:
The camera caches the images internally (Distance image gets calculated from Point3fImage which in turn gets calculated from DepthRawImage). However when the user requests a FloatCameraImage and turns it into a FloatImage the original FloatCameraImage gets destroyed by the constructor of FloatImage. The second time the same channel (or a channel that gets calculated from the former) is requested an invalid FloatCameraImage is returned.
So should we cache the images at all? Or cache them and only return a copy every time? Question is how much more expensive is it to calculate the image compared to copying it.

File: MetriCam2/Camera.cs [hunk 1]
```diff
@@ -1555,6 +1560,12 @@ public CameraImage CalcChannel(string channelName)
             //    return calcChannelCache[channelName];
             //}
 
+            if(imageCache.ContainsKey(channelName))
+            {
+                imageCache.TryGetValue(channelName, out CameraImage cachedImage);
```
Comment (f00f) [inline MetriCam2/Camera.cs]: Is there a reason to use TryGetValue? I think `return imageCache[channelName]` is sufficient.

File: MetriCam2/Camera.cs [hunk 2]
```diff
@@ -1583,6 +1594,7 @@ public CameraImage CalcChannel(string channelName)
                 }
 
                 img = CalcChannelImpl(channelName);
+                imageCache.Add(channelName, img);
```
Comment (f00f) [inline MetriCam2/Camera.cs]: Image is added to cache only in this branch. I would add it right before the `return` statement.

**Comment by f00f** [comment]
I guess Connect or Disconnect should also clear the cache, what do you think?

**Comment by f00f** [comment]
Regarding your question "should we cache at all" and the concern that the cached image might get destroyed by handing out a reference to it, this is what the MetriCam 2 Developer Documentation has to say about it:

* CalcChannelImpl should compute the actual resulting image and cache the result.
* CalcChannelImpl should always return a copy of the data.

**Comment by f00f** [comment]
Regarding which version is faster: I think it's faster to do a block-wise copy than iterate over the same amount of memory _and_ do operations on it.

**Pull request closed without merging**